code/Player/RoleDistributor.cs

```diff
@@ -0,0 +1,214 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Sandbox;
+using ssl.Player.Roles;
+
+namespace ssl.Player
+{
+    public class RoleDistributor
+    {
+        private HashSet<MainPlayer> players;
+        public Scenario Scenario;
+
+        public RoleDistributor(Scenario scenario, HashSet<MainPlayer> players)
+        {
+            this.Scenario = scenario;
+            this.players = players;
+            Log.Info("Remove role for " + this.players.Count);
+            foreach (MainPlayer mainPlayer in this.players)
+            {
+                mainPlayer.RoleHandler.AssignRole(null);
+            }
+        }
+
+        public void Distribute()
+        {
+            Log.Info($"[RoleDistributor] Starting to distribute roles for {players.Count} players..");
+            //Get constraints
+            List<ScenarioConstraint> constraints = Scenario.GetScenarioConstraint(players.Count);
+            Log.Info($"[RoleDistributor] {constraints.Count} constraints finded");
+            //Fulfill constraits
+            foreach (ScenarioConstraint constraint in constraints)
+            {
+                Log.Info($"[RoleDistributor] Constraint {constraint} have to be fullfilled");
+                while (!ConstraintFullfilled(constraint))
+                {
+                    if (FulfillConstraint(constraint))
+                    {
+                        Log.Info($"[RoleDistributor] Constraint {constraint} is fulfilled");
+                    }
+                    else
+                    {
+                        Log.Warning($"[RoleDistributor] Constraint {constraint} is not fulfilled, something wrong");
+                        break;
+                    }
+                }
```

Comment:
This while loop seems useless

Is it equivalent to
while(FulfillConstraint(constraint))
since you break if it's false ?

Reply:
Mhhh you check if the constraints are fulfilled, if not, your try to fulfill. In the case of constraint are not fulfilled and you cant fulfill you get Log.Warning

Reply:
An if should be enough ?
```suggestion
                Log.Info($"[RoleDistributor] Constraint {constraint} have to be fullfilled");
                if (!ConstraintFullfilled(constraint))
                {
                    if (FulfillConstraint(constraint))
                    {
                        Log.Info($"[RoleDistributor] Constraint {constraint} is fulfilled");
                    }
                    else
                    {
                        Log.Warning($"[RoleDistributor] Constraint {constraint} is not fulfilled, something wrong");
                    }
                }
```

Reply:
No because we can have 10 constraints to fulfill

Reply:
But you call FulfillConstraint on the same constraint, the 10 constraints come from the foreach before aren't they ?